src/Avalonia.Base/Interactivity/Interactive.cs

```diff
@@ -114,17 +112,129 @@ public void RemoveHandler<TEventArgs>(RoutedEvent<TEventArgs> routedEvent, Event
         /// <param name="e">The event args.</param>
         public void RaiseEvent(RoutedEventArgs e)
         {
-            e = e ?? throw new ArgumentNullException(nameof(e));
+            ThrowHelper.ThrowIfNull(e, nameof(e));
 
             if (e.RoutedEvent == null)
             {
-                throw new ArgumentException("Cannot raise an event whose RoutedEvent is null.");
+                throw new ArgumentException($"Cannot raise an event whose {nameof(RoutedEventArgs.RoutedEvent)} is null.");
             }
 
-            using var route = BuildEventRoute(e.RoutedEvent);
-            route.RaiseEvent(this, e);
+            using var route = BuildLightweightEventRoute(e.RoutedEvent);
+            route.RaiseEventWithArgs(this, e);
+        }
+
+        /// <summary>
+        /// Raises a routed event by creating the <see cref="RoutedEventArgs"/> only if they're needed.
+        /// Prefer this method over <see cref="RaiseEvent(RoutedEventArgs)"/> to avoid allocations if there's no event listeners.
+        /// </summary>
+        /// <typeparam name="TContext">The arbitrary context type, used to create the event arguments.</typeparam>
+        /// <param name="routedEvent">The routed event.</param>
+        /// <param name="eventArgsFactory">A factory used to create the <see cref="RoutedEventArgs"/>.</param>
+        /// <param name="context">
+        /// An arbitrary object used as a context to create the event arguments.
+        /// Use this argument to avoid closures.
+        /// </param>
+        /// <returns>
+        /// The created <see cref="RoutedEventArgs"/> used to raise the event,
+        /// or null if the event wasn't raised because it didn't have any listeners.
+        /// </returns>
+        public RoutedEventArgs? RaiseEvent<TContext>(
```

Comment:
I wonder if passing the context by value combined with a delegate call is less expensive than dealing with a single allocation

Reply:
It's optimizing for the case where there are no listeners, where it's a clear win.

For the case when there are listeners, passing the state and calling the delegate is a little slower than a direct call, as you'd expect.

Note that the delegate+state pattern appears in .NET itself, for example in [`ILogger.Log<TState>()`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.ilogger.log?view=net-8.0) or [`String.Create<TState>()`](https://learn.microsoft.com/en-us/dotnet/api/system.string.create?view=net-8.0#system-string-create-1(system-int32-0-system-buffers-spanaction((system-char-0)))).



## Benchmarks

### SmallEvent

|                  Method |       Mean |     Error |    StdDev |   Gen0 | Allocated |
|------------------------ |-----------:|----------:|----------:|-------:|----------:|
|   Standard_No_Listeners |  22.091 ns | 0.1672 ns | 0.1564 ns | 0.0024 |      40 B |
|    Factory_No_Listeners |   9.617 ns | 0.0545 ns | 0.0510 ns |      - |         - |
| Standard_With_Listeners | 100.925 ns | 0.8585 ns | 0.8031 ns | 0.0057 |      96 B |
|  Factory_With_Listeners | 111.128 ns | 0.6346 ns | 0.5936 ns | 0.0057 |      96 B |

### LargeEvent:

|                  Method |      Mean |    Error |   StdDev |   Gen0 | Allocated |
|------------------------ |----------:|---------:|---------:|-------:|----------:|
|   Standard_No_Listeners |  39.63 ns | 0.768 ns | 0.854 ns | 0.0076 |     128 B |
|    Factory_No_Listeners |  21.99 ns | 0.115 ns | 0.108 ns |      - |         - |
| Standard_With_Listeners | 118.76 ns | 1.438 ns | 1.345 ns | 0.0110 |     184 B |
|  Factory_With_Listeners | 126.16 ns | 0.411 ns | 0.343 ns | 0.0110 |     184 B |


<details>
<summary>Benchmark code</summary>

#nullable enable

using Avalonia.Controls;
using Avalonia.Input;
using Avalonia.Interactivity;
using BenchmarkDotNet.Attributes;

namespace Avalonia.Benchmarks;

[MemoryDiagnoser]
public class RaiseEvent_SmallEvent_Benchmarks
{
    private static readonly RoutedEvent<RoutedEventArgs> _smallEvent =
        RoutedEvent.Register<Button, RoutedEventArgs>("SomeEvent", RoutingStrategies.Bubble);

    private readonly Button _button = CreateButton(10);

    internal static Button CreateButton(int depth)
    {
        var button = new Button();

        for (int i = 0; i < depth; i++)
        {
            var child = new Button();
            button.Content = child;
            button = child;
        }

        return button;
    }

    [GlobalSetup(Targets = [nameof(Standard_With_Listeners), nameof(Factory_With_Listeners)])]
    public void SetupListeners()
        => _button.AddHandler(_smallEvent, (_, _) => { });

    [Benchmark]
    public void Standard_No_Listeners()
        => Standard();

    [Benchmark]
    public void Factory_No_Listeners()
        => Factory();

    [Benchmark]
    public void Standard_With_Listeners()
        => Standard();

    [Benchmark]
    public void Factory_With_Listeners()
        => Factory();

    private void Standard()
        => _button.RaiseEvent(new RoutedEventArgs(_smallEvent));

    private void Factory()
        => _button.RaiseEvent(_smallEvent);
}

[MemoryDiagnoser]
public class RaiseEvent_LargeEvent_Benchmarks
{
    private static readonly RoutedEvent<PointerEventArgs> _largeEvent =
        RoutedEvent.Register<Button, PointerEventArgs>("LargeEvent", RoutingStrategies.Bubble);

    private readonly Button _button = RaiseEvent_SmallEvent_Benchmarks.CreateButton(10);
    private readonly IPointer _pointer = new Pointer(1, PointerType.Mouse, true);

    [GlobalSetup(Targets = [nameof(Standard_With_Listeners), nameof(Factory_With_Listeners)])]
    public void SetupListeners()
        => _button.AddHandler(_largeEvent, (_, _) => { });

    [Benchmark]
    public void Standard_No_Listeners()
        => Standard();

    [Benchmark]
    public void Factory_No_Listeners()
        => Factory();

    [Benchmark]
    public void Standard_With_Listeners()
        => Standard();

    [Benchmark]
    public void Factory_With_Listeners()
        => Factory();

    private void Standard()
        => _button.RaiseEvent(new PointerEventArgs(
            _largeEvent,
            _button,
            _pointer,
            _button,
            new Point(123.0, 456.0),
            9876543210UL,
            new PointerPointProperties(RawInputModifiers.Control, PointerUpdateKind.LeftButtonPressed),
            KeyModifiers.Control));

    private void Factory()
        => _button.RaiseEvent(
            _largeEvent,
            static (e, ctx) => new PointerEventArgs(e, ctx.source, ctx.pointer, ctx.rootVisual, ctx.rootVisualPosition, ctx.timestamp, ctx.properties, ctx.modifiers),
            (source: _button,
                pointer: _pointer,
                rootVisual: _button,
                rootVisualPosition: new Point(123.0, 456.0),
                timestamp: 9876543210UL,
                properties: new PointerPointProperties(RawInputModifiers.Control, PointerUpdateKind.LeftButtonPressed),
                modifiers: KeyModifiers.Control));
}

</details>

So it's about 6-10% slower when there's listeners, and twice as fast when there's none.
Note that the handler does nothing in these benchmarks - real code will probably have heavier handlers.

If the overhead for listeners isn't acceptable, we can avoid the delegate call and context and use the `LightweightEventRoute` directly in the solution - it's simply less convenient to write than then `RaiseEvent()` handlers.